keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+
+* The CSI driver doesn't support online expansions.
+  * This is a scenario where the driver cannot do an expansion while the volume is published to a node/the pod that's using it is up. When this happens, the CSI driver will return an error when the CSI resizer calls it (eg `FailedPrecondition`). In this proposal we won't be supporting the scenario of not supporting online expansions. Ideally, we'd want to know ahead of time (during StatefulSet admission/validation) if the CSI driver supports online expansion or not - but it might not be so simple. We'll discuss it in details when we talk about [Validation](#validation).
+* The CSI driver doesn't support offline expansions.
+  * This is a scenario where the volume must be published to a node/the pod that's using it must be up in order to perform an expansion. This is not exactly an error. The expansion process will be stuck until the volume is published on a node. This could happen in scenarios where a user attempts to resize a PVC that's not currently attached to a pod (eg a PVC is bound to a pod of the StatefulSet and then the user downscales the StatefulSet). We'll have to make sure to not get in the way of the user when try try to recover from this scenario.
+* No space left/exceeding quota
+  * This is a scenario where the driver returns an error to the resizer (eg `ResourceExhaused`/`OutOfRange`) because there's no space left for the resize process, or the user exceeds a quota on the provider they are using for storage. There is nothing we can do in this scenario other than not getting in the way of the user when they try to recover from it, and, continue reconciliation once they apply successful recovery. The recovery could involve steps like
+    * Using the mechanisms that are described in [KEP 1790](../1790-recover-resize-failure/README.md)
+    * Requesting more quota from the provider
+  * It's important to note that, it's possible for partial failures to occur as well - which might be more tricky for the user to handle. For example, a user could do a resize in a StatefulSet that has 3 replicas and the resize could succeed on 2 of them and fail only on the last.
+    * In such case, the user won't be able to apply the mechanisms in [KEP 1790](../1790-recover-resize-failure/README.md) and their only recourse would be to request more quota from the provider, or downsize their StatefulSet (number of replicas).
+
+### Validation
+
+We'd ideally want as much validation to happen as early as possible, ie during StatefulSet admission/validation. In reality, perfect validation during StatefulSet admission might be very hard to achieve as it involves communication with parts that the StatefulSet admission plugin won't have direct access to, like the CSI driver. Nonetheless, we'll discuss all the possible validation that we could potentially be performing during admission whether or not we'll actually end up performing them in the final implementation.
```

Comment:
admission and validation are two separate processes, see https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/. I'm a bit confused when you're using both, since in the StatefulSet case the only validation that's happening is inside the kube-apiserver. Maybe having a glossary at the beginning to clearly denote when you're talking about apiserver validation, when about admission validation?

Reply:
> StatefulSet admission plugin won't have direct access to

there's no such thing, StatefulSet validation happens within kube-apiserver, only. There can be additional validation, but those are optional.